Ability to message using the connect messaging channel | Phase 2

SmittieC opened the pull request:
I might break this PR up into two after code rabbit reviewed it

## Description
<!-- A summary of the change, the reason for its implementation, and relevent links. 
Include technical details required to understand the change. -->
This completes all the work for the integration, except ability to link a connect messaging channel to your experiment through the UI.
- Bot is able to initiate a conversation based on the prompt given through the API
- Message encryption and decryption
- Tests

## User Impact
<!-- Describe the impact of this change on the end-users. -->

### Demo
<!-- If relevent, include screenshots or a loom video to demonstrate the new behavior
**Include step-by-step instructions to enable functionality of the change-->

### Docs
<!--Link to documentation that has been updated.-->


## Still todo
- Ability to link a connect messaging channel to an experiment
- Manual testing

SmittieC commented:
@coderabbitai full review

snopoke commented on apps/channels/views.py:
```diff
@@ -124,3 +125,11 @@ def new_api_message(request, experiment_id: uuid):
     return Response(data={"response": response})
+
+
+@require_POST
+@csrf_exempt
+@verify_hmac
+def new_connect_message(request: HttpRequest):
+    tasks.handle_connect_messaging_message.delay(json.loads(request.body))
+    return HttpResponse()
```
+1 for some synchronous validation

snopoke commented on apps/chat/channels.py:
```diff
@@ -769,11 +798,10 @@ def send_text_to_user(self, text: str):
+            raise ChannelException("WebChannel requires an existing session")
 
 
 class ConnectMessagingChannel(ChannelBase):
```
what about adding a consent check here somehwere?

snopoke commented on requirements/requirements.in:
```diff
@@ -61,4 +61,5 @@ phonenumberslite
-urllib3>=2.3.0
\ No newline at end of file
+urllib3>=2.3.0
+pycryptodome
```
What is this used for?

snopoke commented on apps/channels/tasks.py:
```diff
@@ -146,3 +156,40 @@ def handle_api_message(
+
+
+def handle_connect_messaging_message(payload: NewMessagePayload):
+    connect_channel_id = payload.get("channel_id")
+
+    try:
+        participant_data = ParticipantData.objects.prefetch_related("participant").get(
+            content_type=ContentType.objects.get_for_model(Experiment), system_metadata__channel_id=connect_channel_id
+        )
+
+        experiment_channel = ExperimentChannel.objects.prefetch_related("experiment").get(
+            platform=ChannelPlatform.CONNECT_MESSAGING, experiment__id=participant_data.object_id
+        )
+    except ParticipantData.DoesNotExist:
+        log.error(f"No participant data found for channel_id: {connect_channel_id}")
+        return
+    except ExperimentChannel.DoesNotExist:
+        log.error(f"No experiment channel found for participant channel_id: {connect_channel_id}")
+        return
```
You could move this to the api view and pass in the `participant_data.id` and `experiment_channel.id` so that more validation happens in the view which is synchronous.

snopoke commented on apps/channels/tasks.py:
```diff
@@ -146,3 +156,40 @@ def handle_api_message(
     return channel.new_user_message(message)
+
+
+def handle_connect_messaging_message(payload: NewMessagePayload):
```
missing celery task decorator

SmittieC commented on apps/experiments/models.py:
```diff
@@ -1389,6 +1390,10 @@ class ParticipantData(BaseTeamModel):
         models.CharField(max_length=255, blank=True, help_text="The base64 encoded encryption key")
     )
 
+    def get_encryption_key_bytes(self):
+        # TODO: What if we use a getter/setting to ensure what we have is always bytes?
+        return base64.b64decode(self.encryption_key)
```
We generate the base64 so we know it will always be valid

SmittieC commented:
@snopoke I addressed all comments in [this PR](https://github.com/dimagi/open-chat-studio/pull/1041) and am closing this one, since I rebased and want a clean slate.
